modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,403 @@
+        } else {
+          [ severity, eventType, message, data ] = args;
+        }
+
+        if (typeof eventType !== 'string') {
+          throw new Error('Event type must be a string');
+        }
+        if (_.isPlainObject(message)) {
+          data = message;
+          message = undefined;
+        }
+        data = data ? { ...data } : {};
+        const aposModule = moduleSelf.__meta?.name ?? '__unknown__';
+
+        if (typeof message === 'string' && message.trim().length > 0) {
+          message = aposModule
+            ? `${aposModule}: ${eventType}: ${message}`
+            : `${eventType}: ${message}`;
+        } else {
+          message = aposModule
+            ? `${aposModule}: ${eventType}`
+            : eventType;
+        }
+        if (self.options.messageAs) {
+          data[self.options.messageAs] = message;
+        }
+        data.module = aposModule;
+        data.type = eventType;
+        data.severity = severity;
+
+        data = self.processRequestData(req, data);
+
+        return [ message, data ];
+      },
+
+      // Enrich the `data` argument with additional information from the request.
+      processRequestData(req, data) {
+        if (!req) {
+          return data;
+        }
+
+        // Keep the proper property order for better readability.
+        const message = {};
+        if (self.options.messageAs) {
+          message[self.options.messageAs] = data[self.options.messageAs];
+        }
+        const {
+          module: _module, type, severity, ...rest
+        } = data;
+
+        return {
+          ...message,
+          module: _module,
+          type,
+          severity,
+          // https://expressjs.com/en/api.html#req.originalUrl
+          url: req.originalUrl,
+          path: req.path,
+          method: req.method,
+          ip: req.ip,
+          query: req.query,
+          requestId: req.requestId || self.apos.util.generateId(),
+          ...rest
+        };
+      },
+
+      // Assess the module filter configuration and determine if the log
+      // should be kept or rejected.
+      //
+      // `moduleSelf` and `args` arguments should be the same as
+      // passed to `processLogArgs(...)`.
+      //
+      // The module and global configs are merged.
+      // The logic is as follows:
+      // - if severity is matched, keep the log (no matter the event type)
+      // - if eventType is matched, keep the log (no matter the severity)
+      shouldKeepEntry(moduleSelf, ...args) {
+        // Detect severity and eventType from the arguments.
+        let severity;
+        let req;
+        let eventType;
+        if (args[1] && typeof args[1].t === 'function') {
+          // eslint-disable-next-line no-unused-vars
+          [ severity, req, eventType ] = args;
+        } else {
+          [ severity, eventType ] = args;
+        }
+        const module = moduleSelf.__meta?.name ?? '__unknown__';
```

Comment:
We only really care about `module`, `severity` and `eventType`, and the same queries are going to be made a lot.

So after this line, check a cache based on this tuple before doing a lot of spreading and Set instantiation.

Reply:
Nice one!